### ISSUE TITLE
Ability for pillar to read other pillar values

### ISSUE THREAD

**Issue by alienzrcoming** [saltstack/salt#6955]
Hi, I have a number of states that could be greatly simplified if pillar could read other pillar values.  One use case example is I have a "walt_warning" pillar value which i populate in all managed files.  I have several pillar values that declare the entire contents of files which are referenced by jinja templates, and instead of having to populate each instances of this with the contents of the salt_warning value, it'd be much easier to just list it as {{ pillar['salt_warning'] }} in the pillar value itself.

Another example would be a situation where pillar declares values that are dependent on another pillar variable the declares the environment a system is in (eg test, staging, live), which is also declared in pillar.

Thanks in advance!

**Comment by uvsmtid** [comment]
### Preserve order ###

I thought it already works by preserving order in which pillar files are loaded.
I was wrong, but the following could be a generic solution.

The approach is simple:
*    Use `include` key in pillar files to declare dependency of content within pilar file on content from another pillar file. It's probably similar to how state's `include` works.
*    In "single pillar file" case, pillar keys are loaded from the top to the bottom of the file (just like definitions in most of the programming languages).

### Pillar example ###

```
# top.sls
base:
    '*':
        - two
```

```
# one.sls
the_first: 111
```

```
# two.sls
include:
    - one
the_second: {{ pillar['the_first'] }}
```

### Problem confirmed ###

Currently `two.sls` fails to compile because it is pillar file and it uses value defined in `pillar`:
```
2015-04-06 22:58:04,519 [salt.pillar      ][CRITICAL] Rendering SLS 'two' failed, render error:
Jinja variable 'dict object' has no attribute 'the_first'
```

### Side notes ###

I don't know what strings it may pull and how complicated it may get to implement, but it's clean and expected behaviour from user point of view.

Issue #12161 might solve it (there is a related pull request #17284).

**Comment by 0xf10e** [comment]
I've first thought ``ext_pillar_first=True`` would make this work by injecting them before the parsing of the rest of the pillar-files begins.

Something like this might even be an option: Either accumulate pillar-data (first process elements of the list ``ext_pillars``, then elements of ``pillar_roots``) or provide access to external pillars via a variable ``ext_pillar`` during processing of ``pillar_roots`` (as ``ext_pillar_first=True`` gives them a higher priority than ``pillar_roots`` anyway).

**Comment by mspinassi** [comment]
I've done this using Puppet + hiera before, and was a tremendous win.
For example, I'd like to have a file with passwords (only one file with passwords let me easily encrypt it before pushing to git), and then in any other pillar file reference those values.

**Comment by uvsmtid** [comment]
There is a solution!
Recent updates in [this issue][1] pointed at loading variables by Jinja itself (which means ability to data in one Jinja file from another).

For example (using :

*   `/srv/pillar/path/to/some_paramters.jinja` - this file (or many like this) under pillar directory which contains loadable variables imported by Jinja elsewhere later:
    ```
    {% load_yaml as vars %}
    regular_yaml:
        with_arbitrary:
            structure:
                - a
                - b
                - c
    {% endload %}
    ```
*   `/srv/pillar/path/to/some_pillar.sls` - this is the pillar written in Jinja which loads variables from Jinja file above:
    ```
    {% from 'path/to/some_paramters.jinja' import vars %}
    variables_loaded_from_another_file: {{ vars }}
    ```

It's not exactly accessing pillar values (because real pillar values are result of rendering).
But it solves problems of pillar files parameterization:
* data duplication (defining the same parameters in more than one file)
* unnecessarily distributed data like using `grains` instead (which are minion-specific)
* avoiding use of variables from Salt configuration as [described here][2]
* keep all configuration parameters centralized in pillar

[1]: https://github.com/saltstack/salt/issues/4348#issuecomment-110348391
[2]: https://github.com/saltstack/salt/issues/12916#issuecomment-104978939

**Comment by mitar** [comment]
I was also hit by this. It seems all pillars are combined together and ordered based on alphabetical order. So I can access those which top-level field are before in alphabetical order, but not later. And this is despite me specifying the order in `top.sls`. I think order in `top.sls` should be respected.

**Comment by mitar** [comment]
I take back. It does not seem to be by alphabetical order.

**Comment by bbinet** [comment]
FYI, I've created the PillarStack `ext_pillar` which supports to read other pillar data.
See https://github.com/bbinet/pillarstack for more information.

**Comment by morsik** [comment]
@uvsmtid: I found another (easier?) solution.

Leave original file as it was:
```yaml
regular_yaml:
  with_arbitrary:
    structure:
      - a
      - b
      - c
```

In file you want to put include:
```yaml
{% import_yaml 'path/to/some_paramters.sls' as vars %}
variables_loaded_from_another_file: {{ vars }}
```

`import_yaml` is documented here: http://docs.saltstack.com/en/latest/ref/renderers/all/salt.renderers.jinja.html#salt.utils.jinja.SerializerExtension

**Comment by uvsmtid** [comment]
@morsik: Yes, it's less cumbersome!

The only caution for those who employ examples above is that `import_yaml` expects **pure YAML only**. In other words, `*.sls` extension is deceiving (as it implies that imported file can be of any format while it must be pure YAML):
```
{% import_yaml 'path/to/some_paramters.sls' as vars %}
```

The [same documentation][1] has examples of `import`-ing Jinja-level variables and macros from one Jinja-template into another. Note the difference of lower parser-level (`import_yaml` or `import_json`) and higher template-level (`import` variables and macros from Jinja code).

## Bottom line ##

Jinja + pure YAML (or JSON) provide decent workaround to the requested feature, but it **does not extend on any SLS renderer**.

[1]: http://docs.saltstack.com/en/latest/ref/renderers/all/salt.renderers.jinja.html

**Comment by 0xf10e** [comment]
Maybe using [libucl](https://github.com/vstakhov/libucl) (BTW: [very nice presentation about UCL in FreeBSD](youtu.be/8l6bhKIDecg)) would mitigate some of the problems. The parser can handle things like includes and reuse of assigned variables on its own.
Not as pythonic as YAML but less of a hassle to write than JSON.

**Comment by DanyC97** [comment]
+1 on proper implementation

**Comment by J-Fricke** [comment]
Found this trying to set some variables in a pillar file and the import_yaml suggestion by @morsik fits the bill perfectly in my use case. 

After reading the comment by @uvsmtid, I simply named my yaml file with .yaml extension to avoid confusion.

**Comment by DanyC97** [comment]
@jfindlay i see this feature is been marked as P1, should i understand the spot lights are turned on and something will come out of it soon?

shall i hope will make it in next release and not in Carbon (which could be close to summer i suspect :) )

**Comment by jfindlay** [comment]
@DanyC97, yes, although I cannot say when this particular feature will be implemented.

**Comment by basepi** [comment]
@DanyC97 Based on current workloads and committed customer features/bugs for Boron, this will likely not be completed in time for Boron. That could very well change, but I wanted to manage expectations.

**Comment by chekolyn** [comment]
If you want at workaround for now take a look at @bbinet comment.
I think pillarstack will do for me at the moment.

**Comment by AndrewPashkin** [comment]
@jberkus 
> My implementation idea is simple: load the pillars in the order given in pillar/top.sls. If a user wants a pillar to refer to another pillar, they need to get the ordering right themselves.

Or expose a function for loading pillars, so user would call it `salt['pillar.load_pillar']('my_variables.sls')`.

**Comment by ketzacoatl** [comment]
FWIW, reclass handles this use case _very_ gracefully, and I would recommend considering reclass if you need some keys to reference other keys.

**Comment by hemebond** [comment]
Do we have an example of what kind of output this feature should generate? Functional specs? Each issue asking for this feature is slightly different.

I did some testing to replicate a simple system I had in a previous (homemade) config system: https://gist.github.com/hemebond/3b31f3ed5ba89dfc0af7655a498b8d2b

**Comment by ketzacoatl** [comment]
FWIW, reclass does this _perfectly_, and has exactly what the primary use cases want out of the general idea. http://reclass.pantsfullofunix.net/

**Comment by hemebond** [comment]
I thought ReClass only interpolated variables that appeared higher in the hierarchy. Is that not the case?

**Comment by AndrewPashkin** [comment]
@jberkus   
> My implementation idea is simple: load the pillars in the order given in pillar/top.sls. If a user wants a pillar to refer to another pillar, they need to get the ordering right themselves.

How, in your opinion, Salt should handle pillar data, provided from a CLI (a.k.a pillar overrides, which are controlled by `pillar_roots_override_ext_pillar` setting)?

**Comment by jberkus** [comment]
@AndrewPashkin 

Beats me, I never got that complex.  My Salt setups were relatively simple; I just joined this issue because I wanted a lot of re-use of variables.  My ordering suggestion was a way to do "baby steps", and otherwise resolve an order-of-resolution issue which was blocking solutions.

**Comment by ketzacoatl** [comment]
FWIW, reclass has the code to show us how to construct the tree and walk it.

**Comment by AndrewPashkin** [comment]
@jberkus, yes I understand, but implementation of your idea implies, that order of when CLI pillars and external pillars are loaded, must also be defined.

**Comment by jberkus** [comment]
Makes sense.  Is the ordering currently undefined?

BTW, just so you know: the project which caused me to comment on this issue has been over for more than a year, so I won't be able to test any solution.

**Comment by AndrewPashkin** [comment]
@jberkus, currently external pillars always loaded before regular ones and are reference-able from regular ones (I've mistaken in my past comment,  `pillar_roots_override_ext_pillar` only affects final composition, not order of loading). External pillars are also merged with CLI pillars. And CLI pillars also override all pillars in the end.

**Comment by jberkus** [comment]
Well, sounds like the order of load is defined, then.  It's perfectly OK if that ordering is defined by Salt.  It's just with the standard pillars where ordering is undefined.

**Comment by AndrewPashkin** [comment]
@jberkus, hm, yes you are right, actually.

**Comment by 0xf10e** [comment]
Yeah, @ketzacoatl, [reclass](https://reclass.pantsfullofunix.net/) (GH: [madduck/reclass](https://github.com/madduck/reclass)) is probably the best choice here.
Not sure if adding a dependency on reclass is a good idea though.

**Comment by ketzacoatl** [comment]
Yea, that's a question beyond me, though I don't think it needs to be made an external dependency. Salt could still make use of the code - it's clean, reliable, and robust. As a user, the experience is awesome.

**Comment by bbinet** [comment]
@ketzacoatl @0xf10e FYI saltclass (which is very similar to reclass) is now directly available in salt from version 2018.3.0.
See: https://docs.saltstack.com/en/latest/topics/releases/2018.3.0.html#new-saltclass-pillar-master-tops-modules

**Comment by ketzacoatl** [comment]
oh wow... so this is already done then.. we should close this ticket as complete.. no?

**Comment by 0xf10e** [comment]
Sweet, that's for pointing this out for us, @bbinet!
Don't have the time to follow the development as close as I'd like lately…

**Comment by morsik** [comment]
@perfecto25 this is something I proposed 8 years ago... and if you read this issue, you can find that this feature was added 5 years... that's quite gold shovel prize :P
